File: api/filing/models/foreign_income_model.py
"""Foreign income models."""
from typing import List, Any
from .api_base_model import ApiBaseModel


class ForeignDividendModel(ApiBaseModel):
    """Model for foreign dividend income."""
    foreign_dividend_id: int | None = None
    filing_id: int | None = None
    country_code: str | None = None
    country_name: str | None = None
    description: str | None = None
    period1: float = 0  # Upto 15-Jun
    period2: float = 0  # 16-Jun to 15-Sep
    period3: float = 0  # 16-Sep to 15-Dec
    period4: float = 0  # 16-Dec to 15-Mar
    period5: float = 0  # 16-Mar to 31-Mar
    total_amount: float | None = None
    currency_type: str = "INR"
    created_at: Any | None = None
    updated_at: Any | None = None


class ForeignInterestModel(ApiBaseModel):
    """Model for foreign interest income."""
    foreign_interest_id: int | None = None
    filing_id: int | None = None
    country_code: str | None = None
    country_name: str | None = None
    description: str | None = None
    amount: float = 0.0
    currency_type: str = "INR"
    created_at: Any | None = None
    updated_at: Any | None = None


class Section89AModel(ApiBaseModel):
    """Income from retirement benefit account u/s 89A."""
    section_89a_id: int | None = None
    filing_id: int | None = None
    period1: float = 0  # Upto 15-Jun
    period2: float = 0  # 16-Jun to 15-Sep
    period3: float = 0  # 16-Sep to 15-Dec
    period4: float = 0  # 16-Dec to 15-Mar
    period5: float = 0  # 16-Mar to 31-Mar
    usa_amount: float = 0
    uk_amount: float = 0
    canada_amount: float = 0
    created_at: Any | None = None
    updated_at: Any | None = None


class ForeignIncomeModel(ApiBaseModel):
    """Combined model for all foreign income types."""
    filing_id: int | None = None
    foreign_dividend: List[ForeignDividendModel] = []
    foreign_interest: List[ForeignInterestModel] = []
    section_89a: Section89AModel | None = None

File: web/filing/models/income/foreign-income-model.ts
/**
 * Foreign Income Models
 */

export interface ForeignDividendModel {
  foreignDividendId?: number | null
  filingId?: number | null
  countryCode?: string | null
  countryName?: string | null
  description?: string | null
  period1: number
  period2: number
  period3: number
  period4: number
  period5: number
  totalAmount?: number | null
  currencyType: string
  createdAt?: string | null
  updatedAt?: string | null
}

export interface ForeignInterestModel {
  foreignInterestId?: number | null
  filingId?: number | null
  countryCode?: string | null
  countryName?: string | null
  description?: string | null
  amount: number
  currencyType: string
  createdAt?: string | null
  updatedAt?: string | null
}

export interface Section89AModel {
  section89aId?: number | null
  filingId?: number | null
  period1: number
  period2: number
  period3: number
  period4: number
  period5: number
  usaAmount: number
  ukAmount: number
  canadaAmount: number
  createdAt?: string | null
  updatedAt?: string | null
}

export interface ForeignIncomeModel {
  filingId?: number | null
  foreignDividend: ForeignDividendModel[]
  foreignInterest: ForeignInterestModel[]
  section89a?: Section89AModel | null
}
